File: cameo-api-cpp/src/coms/impl/zmq/SubscriberZmq.cpp
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

#include "SubscriberZmq.h"
#include "Serializer.h"
#include "Server.h"
#include "JSON.h"
#include "../../../base/CancelIdGenerator.h"
#include "../../../base/Messages.h"
#include "../../../base/RequestSocket.h"
#include "../../../base/impl/zmq/ContextZmq.h"
#include "../../../base/Waiting.h"
#include <sstream>

namespace cameo {
namespace coms {

SubscriberZmq::SubscriberZmq() :
	m_publisherPort(0),
	m_synchronizerPort(0),
	m_numberOfSubscribers(0),
	m_instanceId(0),
	m_ended(false),
	m_canceled(false) {
}

SubscriberZmq::~SubscriberZmq() {
}

void SubscriberZmq::init(int publisherPort, int synchronizerPort, const std::string& publisherName, int numberOfSubscribers, application::Instance & instance) {

	m_serverEndpoint = instance.getEndpoint();
	m_publisherName = publisherName;
	m_publisherPort = publisherPort;
	m_synchronizerPort = synchronizerPort;
	m_numberOfSubscribers = numberOfSubscribers;
	m_instanceName = instance.getName();
	m_instanceId = instance.getId();
	m_statusEndpoint = instance.getStatusEndpoint().toString();
	m_ended = false;
	m_canceled = false;

	// Create a socket for publishing.
	ContextZmq* contextImpl = dynamic_cast<ContextZmq *>(application::This::getCom().getContext());
	m_subscriber.reset(new zmq::socket_t(contextImpl->getContext(), ZMQ_SUB));
	m_subscriber->setsockopt(ZMQ_SUBSCRIBE, message::Event::SYNC, std::string(message::Event::SYNC).length());
	m_subscriber->setsockopt(ZMQ_SUBSCRIBE, message::Event::STREAM, std::string(message::Event::STREAM).length());
	m_subscriber->setsockopt(ZMQ_SUBSCRIBE, message::Event::ENDSTREAM, std::string(message::Event::ENDSTREAM).length());
	m_subscriber->setsockopt(ZMQ_SUBSCRIBE, message::Event::CANCEL, std::string(message::Event::CANCEL).length());
	m_subscriber->setsockopt(ZMQ_SUBSCRIBE, message::Event::STATUS, std::string(message::Event::STATUS).length());

	m_subscriber->connect(m_serverEndpoint.withPort(m_publisherPort).toString());

	// We must first bind the cancel publisher before connecting the subscriber.
	std::stringstream cancelEndpoint;

	// We define a unique name.
	cancelEndpoint << "inproc://cancel." << CancelIdGenerator::newId();
	m_cancelEndpoint = cancelEndpoint.str();

	m_cancelPublisher = std::unique_ptr<zmq::socket_t>(new zmq::socket_t(contextImpl->getContext(), ZMQ_PUB));
	m_cancelPublisher->bind(m_cancelEndpoint.c_str());

	m_subscriber->connect(m_cancelEndpoint.c_str());
	m_subscriber->connect(m_statusEndpoint.c_str());

	// Synchronize the subscriber only if the number of subscribers > 0.
	if (m_numberOfSubscribers > 0) {

		// Create a request socket.
		std::unique_ptr<RequestSocket> requestSocket = application::This::getCom().createRequestSocket(m_serverEndpoint.withPort(m_synchronizerPort).toString());

		// Poll subscriber.
		zmq_pollitem_t items[1];
		items[0].socket = static_cast<void *>(*m_subscriber);
		items[0].fd = 0;
		items[0].events = ZMQ_POLLIN;
		items[0].revents = 0;

		while (true) {

			// The subscriber sends SYNC messages to the publisher that returns SYNC message.
			try {
				requestSocket->requestJSON(createSyncRequest());
			}
			catch (const ConnectionTimeout&) {
				// The server is not accessible.
			}

			// Wait for 100ms.
			int rc = zmq::poll(items, 1, 100);
			if (rc != 0) {
				break;
			}
		}

		requestSocket->requestJSON(createSubscribePublisherRequest());
	}
}

const std::string& SubscriberZmq::getPublisherName() const {
	return m_publisherName;
}

const std::string& SubscriberZmq::getInstanceName() const {
	return m_instanceName;
}

int SubscriberZmq::getInstanceId() const {
	return m_instanceId;
}

Endpoint SubscriberZmq::getInstanceEndpoint() const {
	return m_serverEndpoint;
}

bool SubscriberZmq::isEnded() const {
	return m_ended;
}

bool SubscriberZmq::isCanceled() const {
	return m_canceled;
}

std::optional<std::string> SubscriberZmq::receiveBinary() {

	while (true) {
		std::unique_ptr<zmq::message_t> message(new zmq::message_t());
		m_subscriber->recv(message.get());

		std::string response(static_cast<char*>(message->data()), message->size());

		if (response == message::Event::STREAM) {
			message.reset(new zmq::message_t());
			m_subscriber->recv(message.get());
			return std::string(static_cast<char*>(message->data()), message->size());

		} else if (response == message::Event::ENDSTREAM) {
			m_ended = true;
			return {};

		} else if (response == message::Event::CANCEL) {
			m_canceled = true;
			return {};

		} else if (response == message::Event::STATUS) {
			message.reset(new zmq::message_t());
			m_subscriber->recv(message.get());

			// Get the JSON object.
			json::Object status;
			json::parse(status, message.get());

			int id = status[message::StatusEvent::ID].GetInt();

			if (id == m_instanceId) {
				application::State state = status[message::StatusEvent::APPLICATION_STATE].GetInt();

				// test the terminal state
				if (state == application::SUCCESS
					|| state == application::STOPPED
					|| state == application::KILLED
					|| state == application::FAILURE) {
					// Exit because the remote application has terminated.
					return {};
				}
			}
		}
	}

	return {};
}

std::optional<std::string> SubscriberZmq::receive() {
	return receiveBinary();
}

std::optional<std::tuple<std::string, std::string>> SubscriberZmq::receiveTwoBinaryParts() {

	while (true) {
		std::unique_ptr<zmq::message_t> message(new zmq::message_t());
		m_subscriber->recv(message.get());

		std::string response(static_cast<char*>(message->data()), message->size());

		if (response == message::Event::STREAM) {

			std::tuple<std::string, std::string> result;

			message.reset(new zmq::message_t());
			m_subscriber->recv(message.get());
			std::string data1 = std::string(static_cast<char*>(message->data()), message->size());

			message.reset(new zmq::message_t());
			m_subscriber->recv(message.get());
			std::string data2 = std::string(static_cast<char*>(message->data()), message->size());

			return std::make_tuple(data1, data2);

		} else if (response == message::Event::ENDSTREAM) {
			m_ended = true;
			return {};

		} else if (response == message::Event::CANCEL) {
			return {};

		} else if (response == message::Event::STATUS) {
			message.reset(new zmq::message_t());
			m_subscriber->recv(message.get());

			// Get the JSON object.
			json::Object status;
			json::parse(status, message.get());

			int id = status[message::StatusEvent::ID].GetInt();

			if (id == m_instanceId) {
				application::State state = status[message::StatusEvent::APPLICATION_STATE].GetInt();

				// test the terminal state
				if (state == application::SUCCESS
					|| state == application::STOPPED
					|| state == application::KILLED
					|| state == application::FAILURE) {
					// Exit because the remote application has terminated.
					return {};
				}
			}
		}
	}

	return {};
}

void SubscriberZmq::cancel() {

	std::string m_message = message::Event::CANCEL;

	zmq::message_t requestType(m_message.length());
	std::string data(message::Event::CANCEL);
	zmq::message_t requestData(data.length());
	memcpy(requestType.data(), m_message.c_str(), m_message.length());
	memcpy(requestData.data(), data.c_str(), data.length());
	m_cancelPublisher->send(requestType, ZMQ_SNDMORE);
	m_cancelPublisher->send(requestData);
}

std::string SubscriberZmq::createSubscribePublisherRequest() const {

	json::StringObject request;
	request.pushKey(message::TYPE);
	request.pushInt(message::SUBSCRIBE_PUBLISHER_v0);

	return request.toString();
}

}
}

File: java/api/src/main/java/fr/ill/ics/cameo/coms/impl/zmq/SubscriberZmq.java
/*
 * Copyright 2015 Institut Laue-Langevin
 *
 * Licensed under the EUPL, Version 1.1 only (the "License");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * http://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */

package fr.ill.ics.cameo.coms.impl.zmq;

import org.json.simple.JSONObject;

import fr.ill.ics.cameo.Zmq;
import fr.ill.ics.cameo.base.Application;
import fr.ill.ics.cameo.base.CancelIdGenerator;
import fr.ill.ics.cameo.base.This;
import fr.ill.ics.cameo.base.impl.zmq.ContextZmq;
import fr.ill.ics.cameo.coms.impl.SubscriberImpl;
import fr.ill.ics.cameo.messages.JSON;
import fr.ill.ics.cameo.messages.Messages;
import fr.ill.ics.cameo.strings.Endpoint;

public class SubscriberZmq implements SubscriberImpl {
	
	private Zmq.Context context;
	private Zmq.Socket subscriber;
	private String publisherIdentity;
	private String cancelEndpoint;
	private Zmq.Socket cancelPublisher;
	private int appId;
	private boolean ended = false;
	private boolean canceled = false;
	
	public void init(int appId, Endpoint endpoint, Endpoint appStatusEndpoint, String publisherIdentity) {

		this.appId = appId;
		this.publisherIdentity = publisherIdentity;
		
		this.context = ((ContextZmq)This.getCom().getContext()).getContext();
		
		// Create the subscriber
		subscriber = context.createSocket(Zmq.SUB);
		subscriber.connect(endpoint.toString());
		
		// Subcribe to the publisher.
		subscriber.subscribe(publisherIdentity);
		
		// Create an endpoint that should be unique
		cancelEndpoint = "inproc://cancel." + CancelIdGenerator.newId();
		
		// Create a cancel publisher so that it sends the CANCEL message to the status subscriber (connected to 2 publishers)
		cancelPublisher = context.createSocket(Zmq.PUB);
		cancelPublisher.bind(cancelEndpoint);

		// Subscribe to CANCEL
		subscriber.connect(cancelEndpoint);
		subscriber.subscribe(Messages.Event.CANCEL);
		
		// Subscribe to STATUS
		subscriber.connect(appStatusEndpoint.toString());
		subscriber.subscribe(Messages.Event.STATUS);
	}
	
	public boolean isEnded() {
		return ended;
	}
	
	public boolean isCanceled() {
		return canceled;
	}
	
	/**
	 * 
	 * @return the byte[] data. If the return value is null, then the stream is finished. 
	 */
	public byte[] receive() {

		while (true) {
			String message = subscriber.recvStr();
			
			if (message.equals(publisherIdentity)) {
				
				String messageTypePart = subscriber.recvStr();
				
				// Get the JSON object.
				JSONObject messageType = This.getCom().parse(messageTypePart);
				
				// Get the type.
				long type = JSON.getLong(messageType, Messages.TYPE);
				
				if (type == Messages.STREAM) {
					return subscriber.recv();
				}
				else if (type == Messages.STREAM_END) {
					ended = true;
					return null;	
				}
			}
			else if (message.equals(Messages.Event.CANCEL)) {
				canceled = true;
				return null;
			}
			else if (message.equals(Messages.Event.STATUS)) {
				byte[] statusMessage = subscriber.recv();
				
				// Get the JSON object.
				JSONObject status = This.getCom().parse(statusMessage);
				
				// Get the id.
				int id = JSON.getInt(status, Messages.StatusEvent.ID);
									
				if (appId == id) {
					
					// Get the state.
					int state = JSON.getInt(status, Messages.StatusEvent.APPLICATION_STATE);
					
					// Test if the state is terminal
					if (state == Application.State.SUCCESS 
							|| state == Application.State.STOPPED
							|| state == Application.State.KILLED
							|| state == Application.State.ERROR) {
						// Exit because the remote application has terminated.
						return null;
					}
				}
			}
		}
	}
	
	/**
	 * 
	 * @return the byte[] data. If the return value is null, then the stream is finished. 
	 */
	public byte[][] receiveTwoParts() {

		while (true) {
			String message = subscriber.recvStr();
			
			if (message.equals(publisherIdentity)) {
				
				String messageTypePart = subscriber.recvStr();
				
				// Get the JSON object.
				JSONObject messageType = This.getCom().parse(messageTypePart);
				
				// Get the type.
				long type = JSON.getLong(messageType, Messages.TYPE);
				
				if (type == Messages.STREAM) {
					byte[][] result = new byte[2][];
					result[0] = subscriber.recv();
					result[1] = subscriber.recv();
					
					return result;
				}
				else if (type == Messages.STREAM_END) {
					ended = true;
					return null;	
				}
			}
			else if (message.equals(Messages.Event.CANCEL)) {
				canceled = true;
				return null;
			}
			else if (message.equals(Messages.Event.STATUS)) {
				byte[] statusMessage = subscriber.recv();
				
				// Get the JSON request object.
				JSONObject request =  This.getCom().parse(statusMessage);
				
				// Get the id.
				int id = JSON.getInt(request, Messages.StatusEvent.ID);
				
				if (appId == id) {
					
					// Get the state.
					int state = JSON.getInt(request, Messages.StatusEvent.APPLICATION_STATE);
					
					// Test if the state is terminal
					if (state == Application.State.SUCCESS 
							|| state == Application.State.STOPPED
							|| state == Application.State.KILLED
							|| state == Application.State.ERROR) {
						// Exit because the remote application has terminated.
						return null;
					}
				}
			}
		}
	}
	
	/**
	 * 
	 * @return the string data. If the return value is null, then the stream is finished. 
	 */
	public String receiveString() {
		
		byte[] data = receive();
		
		if (data == null) {
			return null;
		}
		
		return Messages.parseString(data);
	}
	
	public void cancel() {	
	
		cancelPublisher.sendMore(Messages.Event.CANCEL);
		cancelPublisher.send(Messages.Event.CANCEL);
	}
	
	public void terminate() {
		context.destroySocket(subscriber);
		context.destroySocket(cancelPublisher);
	}
	
}